File: producer-api-go-rest-lambda/internal/repository/business_event_repo.go
package repository

import (
	"context"
	"encoding/json"
	"time"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"
)

// BusinessEventRepository handles database operations for business events
type BusinessEventRepository struct {
	pool *pgxpool.Pool
}

// NewBusinessEventRepository creates a new BusinessEventRepository
func NewBusinessEventRepository(pool *pgxpool.Pool) *BusinessEventRepository {
	return &BusinessEventRepository{pool: pool}
}

// Create inserts a new business event into the database
// If conn is provided, uses that connection (for transactions), otherwise acquires from pool
// Returns DuplicateEventError if event ID already exists
func (r *BusinessEventRepository) Create(
	ctx context.Context,
	eventID string,
	eventName string,
	eventType *string,
	createdDate *time.Time,
	savedDate *time.Time,
	eventData map[string]interface{},
	conn pgx.Tx,
) error {
	// Serialize event data to JSON
	eventDataJSON, err := json.Marshal(eventData)
	if err != nil {
		return err
	}

	query := `
		INSERT INTO business_events (id, event_name, event_type, created_date, saved_date, event_data)
		VALUES ($1, $2, $3, $4, $5, $6)
	`

	var execErr error
	if conn != nil {
		// Use provided transaction connection
		_, execErr = conn.Exec(ctx, query, eventID, eventName, eventType, createdDate, savedDate, eventDataJSON)
	} else {
		// Acquire connection from pool
		_, execErr = r.pool.Exec(ctx, query, eventID, eventName, eventType, createdDate, savedDate, eventDataJSON)
	}

	if execErr != nil {
		// Check if it's a unique violation (duplicate key)
		if pgErr, ok := execErr.(*pgx.PgError); ok && pgErr.Code == "23505" {
			return NewDuplicateEventError(eventID, "")
		}
		return execErr
	}

	return nil
}

File: producer-api-python-rest-lambda-pg/repository/business_event_repo.py
"""Business event repository for database operations."""

import asyncpg
import json
import logging
import asyncio
from typing import Optional
from datetime import datetime
from repository.connection_pool import invalidate_connection_pool

logger = logging.getLogger(__name__)

# Retry configuration
MAX_RETRIES = 3
INITIAL_RETRY_DELAY = 0.1  # 100ms
MAX_RETRY_DELAY = 2.0  # 2 seconds


class DuplicateEventError(Exception):
    """Exception raised when attempting to create an event with an existing ID."""
    def __init__(self, event_id: str, message: str = None):
        self.event_id = event_id
        self.message = message or f"Event with ID '{event_id}' already exists"
        super().__init__(self.message)


class BusinessEventRepository:
    """Repository for business event database operations."""
    
    def __init__(self, pool: asyncpg.Pool):
        """Initialize repository with connection pool."""
        self.pool = pool
    
    def _is_retryable_error(self, error: Exception) -> bool:
        """Check if error is retryable (connection/timeout errors)."""
        error_str = str(error).lower()
        retryable_errors = [
            'connection',
            'timeout',
            'network',
            'unable to connect',
            'connection refused',
            'connection reset',
            'connection timed out',
            'server closed the connection',
            'could not connect',
            'network is unreachable',
            'no route to host',
        ]
        return any(err in error_str for err in retryable_errors)
    
    def _is_pool_invalidation_error(self, error: Exception) -> bool:
        """Check if error indicates connection pool should be invalidated (wrong database/connection)."""
        error_str = str(error).lower()
        # Check for errors that suggest wrong database connection
        invalid_pool_errors = [
            '127.0.0.1',  # Connecting to localhost instead of RDS
            'localhost',
            'cannot assign requested address',  # Connection to wrong endpoint
        ]
        return any(err in error_str for err in invalid_pool_errors)
    
    async def _retry_with_backoff(self, func, *args, **kwargs):
        """Retry function with exponential backoff."""
        delay = INITIAL_RETRY_DELAY
        last_error = None
        
        for attempt in range(MAX_RETRIES):
            try:
                return await func(*args, **kwargs)
            except DuplicateEventError:
                # Don't retry duplicate key errors - raise immediately
                raise
            except Exception as e:
                last_error = e
                
                # If this is a pool invalidation error, invalidate the pool before retrying
                if self._is_pool_invalidation_error(e) and attempt == 0:
                    logger.warning(f"Pool invalidation error detected: {e}. Invalidating connection pool...")
                    await invalidate_connection_pool()
                
                # Check if error is retryable
                if not self._is_retryable_error(e) or attempt == MAX_RETRIES - 1:
                    raise
                
                # Exponential backoff
                logger.warning(
                    f"Retryable error on attempt {attempt + 1}/{MAX_RETRIES}: {e}. "
                    f"Retrying in {delay:.2f}s..."
                )
                await asyncio.sleep(delay)
                delay = min(delay * 2, MAX_RETRY_DELAY)
        
        # If we exhausted retries, raise the last error
        raise last_error
    
    async def create(self, event_id: str, event_name: str, event_type: Optional[str],
                     created_date: Optional[datetime], saved_date: Optional[datetime],
                     event_data: dict, conn: Optional[asyncpg.Connection] = None) -> None:
        """Create a new business event with retry logic.
        
        Args:
            event_id: Event ID
            event_name: Event name
            event_type: Event type
            created_date: Created date
            saved_date: Saved date
            event_data: Full event JSON as dict
            conn: Optional database connection (for transactions). If not provided, acquires from pool.
        
        Raises:
            DuplicateEventError: If an event with the same ID already exists (409 Conflict)
        """
        async def _do_create():
            if conn:
                # Use provided connection directly (transaction managed externally)
                await self._insert_business_event(conn, event_id, event_name, event_type,
                                                 created_date, saved_date, event_data)
            else:
                # Acquire connection from pool
                async with self.pool.acquire() as connection:
                    await self._insert_business_event(connection, event_id, event_name, event_type,
                                                     created_date, saved_date, event_data)
        
        await self._retry_with_backoff(_do_create)
    
    async def _insert_business_event(self, conn: asyncpg.Connection, event_id: str, event_name: str,
                                    event_type: Optional[str], created_date: Optional[datetime],
                                    saved_date: Optional[datetime], event_data: dict,
                                    skip_diagnostics: bool = False) -> None:
        """Insert business event into database.
        
        Args:
            conn: Database connection
            event_id: Event ID
            event_name: Event name
            event_type: Event type
            created_date: Created date
            saved_date: Saved date
            event_data: Event data as dict
            skip_diagnostics: Unused (kept for compatibility)
        """
        try:
            await conn.execute(
                """
                INSERT INTO business_events (id, event_name, event_type, created_date, saved_date, event_data)
                VALUES ($1, $2, $3, $4, $5, $6)
                """,
                event_id,
                event_name,
                event_type,
                created_date,
                saved_date,
                json.dumps(event_data),
            )
        except asyncpg.UniqueViolationError as e:
            # Raise custom exception for duplicate key violations
            logger.warning(f"Duplicate event ID detected: {event_id}")
            raise DuplicateEventError(event_id, f"Event with ID '{event_id}' already exists") from e
